VSS.md

```diff
@@ -0,0 +1,518 @@
+
+Binary RDB and replication stream compatibility with Redisearch is not needed, meaning that online/live data migration will not be possible.
+
+**HASH and JSON Data Type support CUJ** \- As a user, I want to store my vectors in a HASH or JSON data type, which is natively supported in Valkey.
+
+**Scale Up & Out CUJ** \- As a user, I want to be able to both scale up and out
+
+- Requirement: multi-threading for scale up  
+- Requirement: Cluster for scale out & storing more data
+
+**Consistency CUJ** \- As a user, I want to be able to select a consistency model that matches my needs.
+
+- Requirement: A sequential consistency model must be supported.
+
+**Memory footprint CUJ** \- As a user, I want the index to consume memory based on usage and dynamically grow
+
+- Requirement: Lazy memory consumption based on the index entries and the dimensions.  
+- Requirement: Index size should be allowed to grow dynamically with minimal performance overhead
+
+**Replication CUJ** \- As a user, I want to rely on replication to provide high-availability and ensure that my data is protected by storing it across multiple nodes with low replication latency
+
+- Requirement: Support fast bootstrap  
+- Requirement: Support index full sync on replication lag.
+
+**Monitoring CUJ** \- As a user, I need to be able to monitor indexes and re-indexing with simple observability metrics
+
+- Requirement: Ability to monitor index base statistics  
+- Requirement: Ability to monitor cross indexes, aggregated, view statistics.
+
+**Backfill CUJ** \- As a user, I need to be able to create an index after the keyspace contains vector entries
+
+- Requirement: Index creation triggers a backfill process which indexes the existing keyspace entries.
+
+**Hybrid Query CUJ** \- As a user, I want to be able to perform hybrid queries to incorporate filtering criteria in addition to best-match vector search
+
+- Requirement: Support for filtering numeric and tag-based filters.  
+- Requirement: Support for boolean combinations of operators in combination with filtering (AND, OR, NOT)
+
+**Result Set Management CUJ** \- As a user I want to dictate which fields are returned in a search
+
+- Requirement: Support query controls to set the desired returned fields.
+
+**ACLs CUJ** \- As a user of Access Control Lists (ACLs), I want to manage user-based permissions, specific command permissions, etc to manage vector search access
+
+- Requirement: Existing ACls functionality should extend to vector search commands
```

Comment:
I'm missing some details about ACL.

* ACL per command automatically extends to module commands, so we get these automatically, right?
* Does the module create a new ACL category for the VSS commands? (There's module API for that, IIRC.)
* ACL rules per key pattern, do these extend to searches? How/why not?

Reply:
None of the search commands actually have keys on the command, so the existing built-in ACL support for modules doesn't apply. Separately AWS has implemented module extensions for ACLs for search. This will be contributed, but under a separate RFC.

Reply:
ACL is not only about keys. You can allow and disallow commands:

       • +<command>: Add the command to the list of commands the user can call.  Can be used with | for allowing subcommands (e.g “+config|get”).

       • -<command>:  Remove the command to the list of commands the user can call.  Starting Valkey 7.0, it can be used with | for blocking sub‐
         commands (e.g “-config|set”).

Reply:
Right, module-added commands are already supported by ACL. We will flesh out the details of how ACL will work in a separate RFC.